### DISCUSSION TITLE
JAVA 연서, 은진

### DISCUSSION THREAD

**Discussion by watty-exclamationMark** [Shortudy-10th/tech-for-developer#17] [질의 응답]
질의응답!

- 자바소스파일의 컴파일 & 실행과정에 대해 설명해주세요.

- String, StringBuilder, StringBuffer에 대해 설명해주세요.

- 자바의 자료형에 대해 아는 데로 설명해주세요.

- 추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

- 자바 컬렉션에 대해서 설명해주세요.

- 객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.

- 오버라이딩과 오버로딩에 대해 설명해주세요

- call by value / call by reference 에 대해 설명해주세요

- JAVA에는 접근 제어자가 4가지 있습니다.각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

- 자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.

**Comment by watty-exclamationMark** [comment]
### ❓️자바소스파일의 컴파일 & 실행과정에 대해 설명해주세요.

### 💬답변
> 1. 사용자가 소스 코드를 작성한다.
> 2. 컴파일 단계: 자바 소스 코드 파일을 바이트 코드 파일로 변환한다. 만일 문법 오류가 있을 경우 오류 메시지 출력한다.
>     문제가 없다면 해당 클래스의 바이트 코드를 생성, 이를 .class 확장자를 가지는 파일로 저장한다.
> 3. 실행:  컴파일된 바이트 코드 파일(.class)은 JVM에 의해 해당 플랫폼에서 실행 가능한 기계어로 변환된다.
>               이후 JVM은 클래스 로더(Class Loader)를 사용하여 필요한 클래스들을 동적으로 로딩하고, 메모리에 
                  올리고, 프로그램의 진입점인 public static void main(String[] args) 메서드를 찾아 실행한다.

**Comment by watty-exclamationMark** [comment]
### ❓️String, StringBuilder, StringBuffer에 대해 설명해주세요.
### 💬답변
> **String**
>문자열을 표현하는 가장 기본적인 클래스이다.
>문자열을 생성한 후에는 변경할 수 없는 불변 객체이다.
> (예를 들어 String str = “hi”;를 생성한 후 str = str+”JAVA!”;를 실행하면, 새로운 문자열이 생성된 후 str 변수에 할당된다.)
> 
> **StringBuilder**
> String과 다르게 가변한 클래스이다. 즉, 문자열을 동적으로 변경하기 위한 클래스이다.
> (⇒ StringBuilder sb = new StringBuilder(“hi”); 를 생성한 후,  sb.append(”JAVA!”);를 실행하면, 기존 문자열 “hi”에 “JAVA!”가 추가되는 형식이다.)
> StringBuilder의 메서드들은 스레드 동기화를 처리하지 않기 때문에, 단일 스레드 환경에서 사용하는 것이 적합하다.
>
> **StringBuffer**
> StringBuilder처럼 문자열을 동적으로 변경하는데 사용한다.
> 그러나 StringBuilder와 다르게 StringBuffer의 메서드는 물론 내부 데이터 구조 또한 스레드 동기화가 되어 있기 때문에, 멀티 스레드 환경에서 사용하는 것이 적합하다.

**Comment by watty-exclamationMark** [comment]
### **❓️자바의 자료형에 대해 아는 데로 설명해주세요.**

### **💬답변**
> **기본 타입 (Primitive Types)**
> 리터럴을 직접 저장하는 타입이다.
>
> 1. 정수형
>   - byte (8bit)
>   - short (16bit)
>   - int (32 bit)
>   - long (64bit)
> 2. 실수형
>   - float (32bit)
>   - double (64bit)
> 3. 문자형
>   - char (16bit, 유니코드 문자)
> 4. 논리형
>   - boolean (8bit)
>
> **참조 타입 (Reference Types)**
> 객체의 참조를 저장하는 타입으로, Heap 메모리에 객체가 저장된다.


> 요약
> 1. PDT 기본형 : boolean (참 거짓 _-> 논리형_) byte short int long (정수) float double (실수) _char (문자)_
> 2. RDT 참조형 : 배열, 객체 등 PDT를 제외한 모든 DT이 포함된다. 
    _객체의 참조를 저장한다. Heap 메모리에 객체가 저장된다._

**Reply by watty-exclamationMark** [reply]
기본 타입과 참조 타입은 각각 어떤 메모리 영역 위에 올라갈까요?

**Reply by aeroej** [reply]
> Java에서 메모리 영역은 크게 stack과 heap으로 나뉜다.
> 기본 타입은 stack, 참조 타입은 heap 영역에 올라간다.

**Comment by watty-exclamationMark** [comment]
### ❓️추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.
### 💬답변
> 1. 추상 클래스: 다중 상속이 불가능하고, 접근 제어자의 제약이 없다. 
> 2. 인터페이스 : 다중 상속이 가능하다. 선언부가 있는 메서드를 가질 수 없으며 (모두 추상 메서드여야 한다), 변수를 선언하려면 public static final로 선언하여야 한다. 


> **공통점**
> 1. 추상화 : 구현부가 없는 메서드가 존재한다.
> 2. 강제성 : 반드시 구현 해야하는 메서드를 정의하여 강제성을 부여한다.
>
> **차이점**
> 추상 클래스 = 추상 메서드 + 일반 메서드 
> 단일 상속
>
> 인터페이스 = only 추상 메서드 + 상수 -> 자바 8 이후 default, static 메서드 허용
> 다중 상속
> 접근 제어자 : 모두 상수 public static final + 모든 메서드 public
> XXX생성자XXX

**Reply by watty-exclamationMark** [reply]
어떤 때에 추상 클래스를 사용하고 어떤 때에 인터페이스를 사용할까요?

**Reply by aeroej** [reply]
> 추상 클래스 : 공통된 기능이나 상태를 공유해야 할 때. 특정 메서드의 기본 구현을 제공하고 싶을 때.
> 인터페이스 : 공통된 상태나 기본 구현이 필요하지 않을 때. 다중 상속이 필요할 때

**Comment by watty-exclamationMark** [comment]
### ❓️자바 컬렉션에 대해서 설명해주세요.
### 💬답변
> 데이터 관리를 위한 자료구조를 제공하는 프레임워크다.
> List, Queue, HashMap 등의 다양한 형태의 자료구조가 있다.
> 3대 자료구조 인터페이스로 List, Set, Map이 있다.
> Generic 타입을 한정하므로, 형 변환 번거로움이 없다.
> DB처럼 CRUD 기능을 제공한다.

**Comment by watty-exclamationMark** [comment]
### ❓️객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.
### 💬답변
> OOP is A PIE
> 1. 추상화 : 객체의 공통 특징 단순화
> 자식 클래스에서 반드시 메서드 재정의해서 사용하는 경우, 부모 클래스에서의 구현이 무의미하다.
> 구현 강제성 -> 안정성
> 2. 다형성 : 하나의 객체가 여러 형(타입)을 가질 수 있음.
> 3. 상속 : 부모 클래스의 멤버(변수, 메소드)를 자식 클래스가 상속받아 재사용.
> 재사용성, 유지보수성
> 4. 데이터 보호와 은닉. 캡슐화 : 외부로부터 접근 제한. 
> 클래스 멤버변수를 private로 선언하여 데이터 은닉 -> getter setter 메서드로 접근제어

**Comment by watty-exclamationMark** [comment]
### ❓️오버라이딩과 오버로딩에 대해 설명해주세요.
### 💬답변
>1. 오버라이딩: 하위 클래스에서 상위 클래스에 상속받은 메서드를 다시 정의하는 것이다.
> 2. 오버로딩: 동일한 이름의 메서드라도 받는 파라미터의 개수나 type를 다르게 정의하면 다른 메서드이다.

 
> 1. 오버라이딩 : 덮어쓰기! 메서드 재정의.
부모 클래스에서 상속받은 메서드를 자식 클래스에서 재정의하는 것이다.
> 2. 오버로딩 : 추가 적재! 생성자.
파라미터의 타입과 개수를 달리하면 동일한 이름의 메서드를 여러 개 정의할 수 있다.
생성자를 정의할 때 많이 사용한다.
>
> +) 사용자가 별도로 생성자를 정의하면, 자체적으로 만들어져 있는 기본 생성자가 삭제된다.

**Reply by watty-exclamationMark** [reply]
오버로딩 / 오버라이딩 사용시 주의해야 할 점이 무엇일까요? (헷갈릴만한 부분)

**Reply by aeroej** [reply]
> **오버로딩**
> 1. 파라미터의 타입, 개수, 순서 중 적어도 하나는 달라야 한다.
>
> **오버라이딩**
> 1. 부모 클래스의 메서드와 이름, 파라미터가 동일해야 한다.
> 2. 부모 클래스의 접근제어자보다 범위가 같거나 넓어야 한다.
> @Override annotation -> 컴파일러가 실수를 확인해준다.

**Comment by watty-exclamationMark** [comment]
###  ❓️call by value / call by reference 에 대해 설명해주세요.
### 💬답변
> 1. call by value: 값에 의한 호출을 말한다. 매개변수로 그 인자가 가지고 있는 값만을 보내는 것.
> 2. call by reference: 참조에 의한 호출을 말한다. 매개변수로 인자의 참조를 보내어 메모리 내의 원본 변수에 접근하여 값을 바꾼다.
>
> 자바는 call by value 기반의 call by sharing 방식이라고 한다. 
> 객체의 값이 복사되어 함수에 건너지지만 객체를 전달할 때, 객체의 참조 값(주소)도 같이 전달되어 마치 call by reference처럼 보인다고 합니다...
 
> 1. call by value : 파라미터에 PDT (Primitive data type)을 보내는 것
> PDT의 값 자체를 전달한다.
> 2. call by reference : 파라미터에 RDT (Reference data type)을 보내는 것
> 사실상 _객체를 참조하는_ 주소값을 전달한다.

**Comment by watty-exclamationMark** [comment]
### ❓️JAVA에는 접근 제어자가 4가지 있습니다.각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.
### 💬답변
> 1. Private: 해당 클래스 내부에서만 사용이 가능하다.
> 2. Default (아무런 접근 제어자를 명시하지 않았을 때 기본 값이다) : 같은 패키지 내부에서 사용이 가능하다.
> 3. Protected : 같은 패키지 내부, 자식 클래스(= 파생 클래스; 상속받은 클래스)에서 사용이 가능하다.
> 4. Public : 다른 클래스, 다른 패키지에서도 사용이 가능하다.

> 요약
> 1. private : 해당 클래스 내에서만 사용 가능
> 2. default : 해당 패키지 내에서만 사용 가능
> 3. protect : 상속받은 경우에 한해서만 외부 패키지에서 사용 가능
> 4. package : 외부 패키지에서 사용 가능

**Comment by watty-exclamationMark** [comment]
### ❓️자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.
### 💬답변
> 1. extends Thread
> 스레드 클래스를 상속받는 방법이다.
> 2. implements Runnable
> Runnable 인터페이스를 상속받는 방법이다.
>
>차이점 : 자바는 단일 상속만을 지원하기 때문에 Thread 클래스를 상속받는 방식으로 구현하면 다른 클래스를 상속받지 못해 Runnable를 상속받는 방법보단 유연성이 떨어집니다.

**Reply by watty-exclamationMark** [reply]
어떤 때에 Thread를 상속받고 어떤 때에 Runnable 인터페이스를 상속 받는 것이 좋을까요?
혹은 둘 중 어느것을 가장 사용하는 것이 용이할까요? 본인의 생각을 적기!

**Reply by aeroej** [reply]
> 1. extends Thread **클래스**
> 다른 클래스를 상속받을 수 없음 -> 매우매우 강제성이 강하닷!
> 2. implements Runnable **인터페이스**
> 일반적으로 선호하는 방식
> 코드 유연성, 재사용성 높임
> 다중 상속의 제약을 피할 수 있음
